File: Assets/Scripts/CutsceneManager.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using TMPro;

// Manages the intro cutscene.
public class CutsceneManager : MonoBehaviour
{

    // The music manager prefab.
    [SerializeField] GameObject mm_prefab;

    // The text component.
    TextMeshPro tmp;

    // All the messages to be displayed.
    string[] messages = { "The year is 30XX.\nHumans have begun using fusion as their main power source.",
        "Suddenly, leagues of alien robots swarm  Earth. Space pirates!",
        "Having depleted their own fossil fuels, they scour the cosmos looking for planets to raid; Earth is next.",
        "To stop them, the world powers have united to build a ray gun that can destroy the robots.",
        "They put their trust in you to man this machine, and save humanity.",
        "This, soldier, is your Last Assignment: Save Earth from the Robots..." };


    void Start()
    {
        // If played before, skip the cutscene.
        if (PlayerPrefs.GetInt("played") == 1) UnityEngine.SceneManagement.SceneManager.LoadScene("MainMenu");

        // Otherwise:
        else
        {
            // Get the text.
            tmp = FindAnyObjectByType<TextMeshPro>();

            // Instantiate the music manager if it doesn't already exist and turn off the drum layer.
            if (GameObject.FindGameObjectWithTag("music") == null) Instantiate(mm_prefab);
            EventBus.Publish(new MusicEvent("Drum 1", 0.0f));

            // Start the cutscene coroutine.
            StartCoroutine(PlayCutscene());
        }
    }

    // Cutscene coroutine.
    IEnumerator PlayCutscene()
    {

        // For every message:
        for (int i = 0; i < messages.Length; i++)
        {
            // Update the text.
            tmp.text = messages[i];

            // Fade in, wait for a time proportional to the message length, fade out, and briefly pause.
            yield return FadeIn();
            yield return new WaitForSeconds(messages[i].Length / 50.0f);
            yield return FadeOut();
            yield return new WaitForSeconds(0.5f);

        }

        // Load the main menu.
        UnityEngine.SceneManagement.SceneManager.LoadScene("MainMenu");
    }

    // Fade in Coroutine.
    IEnumerator FadeIn()
    {
        // While the text alpha is not full, increase every frame.
        while (tmp.color.a < 1.0f)
        {
            tmp.color = new Color(tmp.color.r, tmp.color.g, tmp.color.b, tmp.color.a + 0.01f);
            yield return new WaitForSeconds(0.0166f);
        }
    }

    // Fade out Coroutine.
    IEnumerator FadeOut()
    {
        // While the text alpha is not full, increase every frame.
        while (tmp.color.a > 0.0f)
        {
            tmp.color = new Color(tmp.color.r, tmp.color.g, tmp.color.b, tmp.color.a - 0.01f);
            yield return new WaitForSeconds(0.0166f);
        }
    }

}

File: component_types/CutsceneManager.lua
CutsceneManager = {

    messages = {{"The year is 30XX.","Humans have begun","using fusion as their","main power source."},
                {"Suddenly, leagues of","alien robots swarm","Earth. Space pirates!"},
                {"Having depleted their","own fossil fuels, they","scour the cosmos","looking for planets to","raid; Earth is next."},
                {"To stop them, the","world powers have","united to build a ray","gun that can destroy","the robots."},
                {"They put their trust","in you to man this","machine, and save","humanity."},
                {"This, soldier, is your","Last Assignment:","Save Earth from the","Robots..."}},
    shows = {110,92,152,131,96,101},
    cur_message = 1,
    cur_mode = 1,
    frame_goal = 0,

    OnStart = function (self)
        self.trs = self.actor:GetComponents("TextRenderer")
        local length = #(self.messages[1])
        for index, value in ipairs(self.messages[1]) do
            self.trs[index].text = value
            self.trs[index].y_position = (length - 1) / 2 + 1 - index
        end
        local played = io.open(Application.FullPath("resources/.data/.played"),"r")
        if played ~= nil then
            io.close(played)
            --Scene.Load("menu")
            --return
        end
        played = io.open(Application.FullPath("resources/.data/.played"),"w")
        io.close(played)
        local sd = Actor.Instantiate("StaticData")
        Actor.DontDestroy(sd)
        sd:GetComponent("MusicManager"):Init()
        Audio.SetEventParameter("event:/music/Theme", "Drum 1", 0);
        self.frame_goal = Application.GetFrame() + 100
    end,

    OnUpdate = function (self)
        local cur_frame = Application.GetFrame()
        if cur_frame == self.frame_goal then
            self.cur_mode = self.cur_mode + 1
            if self.cur_mode == 5 then
                self.cur_message = self.cur_message + 1
                if self.cur_message == 7 then
                    Scene.Load("menu")
                    return
                end
                self.trs[4].text = " "
                self.trs[5].text = " "
                local length = #(self.messages[self.cur_message])
                for index, value in ipairs(self.messages[self.cur_message]) do
                    self.trs[index].text = value
                    self.trs[index].y_position = (length - 1) / 2 + 1 - index
                end
                self.cur_mode = 1
                self.frame_goal = self.frame_goal + 100
            elseif self.cur_mode == 2 then
                self.frame_goal = self.frame_goal + self.shows[self.cur_message]
            elseif self.cur_mode == 3 then
                self.frame_goal = self.frame_goal + 100
            else
                self.frame_goal = self.frame_goal + 30
            end
        else
            if self.cur_mode == 1 then
                for i = 1, 5, 1 do
                    self.trs[i].a = self.trs[i].a + 2.55
                end
            elseif self.cur_mode == 3 then
                for i = 1, 5, 1 do
                    self.trs[i].a = self.trs[i].a - 2.55
                end
            end
        end
    end
}